test/test_builders.go

```diff
@@ -0,0 +1,156 @@
+///usr/bin/true; exec /usr/bin/env go run "$0" "$@"
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+)
+
+func main() {
+	version := os.Args[1]
+
+	os.Setenv("FUNC_REGISTRY", "example.com/jdoe")
+
+	wd, err := os.Getwd()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	packCmd := filepath.Join(wd, "pack")
+
+	funcBinaries := []string {
+		filepath.Join(wd, "bin", "func_snapshot"),
+		filepath.Join(wd, "bin", "func_stable"),
+	}
+	templates := []string {"http", "events" }
+
+	builders := []struct {
+		BuilderImage string
+		Runtime      []string
+	}{
+		{
+			BuilderImage: "quay.io/boson/faas-go-builder",
+			Runtime: []string{"go"},
+		},
+		{
+			BuilderImage: "quay.io/boson/faas-jvm-builder",
+			Runtime: []string{"quarkus", "springboot" },
+		},
+		{
+			BuilderImage: "quay.io/boson/faas-nodejs-builder",
+			Runtime: []string{"node", "typescript"},
+		},
+		{
+			BuilderImage: "quay.io/boson/faas-python-builder",
+			Runtime: []string{"python"},
+		},
+		{
+			BuilderImage: "quay.io/boson/faas-quarkus-native-builder",
+			Runtime: []string{"quarkus"},
+		},
+		{
+			BuilderImage: "quay.io/boson/faas-rust-builder",
+			Runtime: []string{"rust"},
+		},
+	}
+
+	for i := range builders {
+		builders[i].BuilderImage = fmt.Sprintf("%s:%s", builders[i].BuilderImage, version)
+	}
+
+	oldWD, err := os.Getwd()
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer os.Chdir(oldWD)
+
+	testDir, err := os.MkdirTemp("", "test_func")
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer os.RemoveAll(testDir)
+	os.Chdir(testDir)
+
+	var errors []string
+	var fnNumber int
+	for _, funcBinary := range funcBinaries {
+		for _, builder := range builders {
+			for _, runtime := range builder.Runtime {
+				for _, template := range templates {
+
+					fnNumber++
+					func() {
+						defer os.Chdir(testDir)
+
+						fnName := fmt.Sprintf("fnno-%d", fnNumber)
+
+						tp := fmt.Sprintf("test_params{ binary: %q, runtime: %q, template: %q, builder %q }",
+							funcBinary, runtime, template, builder.BuilderImage)
+
+						log.Printf("building function %s", tp)
+
+						cmd := exec.Command(funcBinary,
+							"create", fnName,
+							"--runtime", runtime,
+							"--template", template,
+							"--verbose")
+						cmd.Stdout = os.Stdout
+						cmd.Stderr = os.Stderr
+
+						err := cmd.Run()
+						if err != nil {
+							msg := fmt.Sprintf("failed to create a function (%s): %s", tp, err.Error())
+							log.Println(msg)
+							errors = append(errors, msg)
+							return
+						}
+						os.Chdir(fnName)
+
+						cmd = exec.Command(packCmd,
+							"build", "example.com/jdoe/" + fnName + ":latest",
+							"--builder", builder.BuilderImage,
+							"--pull-policy=if-not-present",
```

Comment:
Should this be `never`? Presumably we only ever want to be testing just-built images, which means they shouldn't be pulled.

Reply:
if you are running not-trusted builder you need to pull `lifecycle` images, maybe I could pull some ahead but I would be sure what tag to pull.

Reply:
@lance  speaking of `lifecycle`: we should figure how to put it into our `quay.io` and pull it from there because there are pull limits on `docker.io` that CI can reach quickly